File: android/app/src/main/java/com/obubba/app/plugins/StorePlugin.java
package com.obubba.app.plugins;

import android.app.Activity;
import android.content.SharedPreferences;
import android.util.Log;
import androidx.annotation.NonNull;
import com.android.billingclient.api.*;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.JSObject;
import com.getcapacitor.JSArray;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@CapacitorPlugin(name = "OBStore")
public class StorePlugin extends Plugin implements PurchasesUpdatedListener {

    private static final String TAG = "OBStore";
    // v1 kept for existing subscribers (Google grandfathers their price)
    private static final List<String> PRODUCT_IDS = Arrays.asList(
        "com.obubba.premium.monthly",
        "com.obubba.premium.annual",
        "com.obubba.premium.lifetime",
        "com.obubba.premium.monthly.v2",
        "com.obubba.premium.annual.v2",
        "com.obubba.premium.lifetime.v2"
    );
    private static final List<String> SUB_IDS = Arrays.asList(
        "com.obubba.premium.monthly",
        "com.obubba.premium.annual",
        "com.obubba.premium.monthly.v2",
        "com.obubba.premium.annual.v2"
    );
    private static final List<String> LIFETIME_IDS = Arrays.asList(
        "com.obubba.premium.lifetime",
        "com.obubba.premium.lifetime.v2"
    );

    private BillingClient billingClient;
    private PluginCall pendingPurchaseCall;
    private List<ProductDetails> cachedProducts = new ArrayList<>();
    private boolean billingConnecting = false;
    private final List<Runnable> billingConnectedQueue = new ArrayList<>();
    private final List<Runnable> billingFailedQueue = new ArrayList<>();

    @Override
    public void load() {
        billingClient = BillingClient.newBuilder(getContext())
            .setListener(this)
            .enablePendingPurchases(
                PendingPurchasesParams.newBuilder()
                    .enableOneTimeProducts()
                    .build()
            )
            .enableAutoServiceReconnection()
            .build();
        connectBilling(null);
    }

    private void connectBilling(Runnable onConnected) {
        connectBilling(onConnected, null);
    }

    private void connectBilling(Runnable onConnected, Runnable onFailed) {
        try {
            if (billingClient.isReady()) {
                if (onConnected != null) onConnected.run();
                return;
            }
            synchronized (this) {
                if (billingConnecting) {
                    if (onConnected != null) billingConnectedQueue.add(onConnected);
                    if (onFailed != null) billingFailedQueue.add(onFailed);
                    return;
                }
                billingConnecting = true;
            }
            billingClient.startConnection(new BillingClientStateListener() {
                @Override
                public void onBillingSetupFinished(@NonNull BillingResult result) {
                    try {
                        List<Runnable> successCallbacks = new ArrayList<>();
                        List<Runnable> failureCallbacks = new ArrayList<>();
                        synchronized (StorePlugin.this) {
                            billingConnecting = false;
                            if (result.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                                if (onConnected != null) successCallbacks.add(onConnected);
                                successCallbacks.addAll(billingConnectedQueue);
                            } else {
                                if (onFailed != null) failureCallbacks.add(onFailed);
                                failureCallbacks.addAll(billingFailedQueue);
                            }
                            billingConnectedQueue.clear();
                            billingFailedQueue.clear();
                        }
                        if (result.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                            Log.i(TAG, "Billing connected");
                            runCallbacks(successCallbacks);
                        } else {
                            Log.w(TAG, "Billing connect failed: " + result.getDebugMessage());
                            runCallbacks(failureCallbacks);
                        }
                    } catch (Exception e) {
                        Log.e(TAG, "Billing setup callback error", e);
                        if (onFailed != null) onFailed.run();
                    }
                }
                @Override
                public void onBillingServiceDisconnected() {
                    Log.w(TAG, "Billing disconnected");
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "Billing connect error", e);
            List<Runnable> failureCallbacks = new ArrayList<>();
            synchronized (this) {
                billingConnecting = false;
                if (onFailed != null) failureCallbacks.add(onFailed);
                failureCallbacks.addAll(billingFailedQueue);
                billingConnectedQueue.clear();
                billingFailedQueue.clear();
            }
            runCallbacks(failureCallbacks);
        }
    }

    private void runCallbacks(List<Runnable> callbacks) {
        for (Runnable callback : callbacks) {
            try {
                if (callback != null) callback.run();
            } catch (Exception e) {
                Log.e(TAG, "Billing callback error", e);
            }
        }
    }

    @PluginMethod
    public void getProducts(PluginCall call) {
        connectBilling(() -> {
            // Query subscriptions
            QueryProductDetailsParams subParams = QueryProductDetailsParams.newBuilder()
                .setProductList(buildProductList(SUB_IDS, BillingClient.ProductType.SUBS))
                .build();

            QueryProductDetailsParams inappParams = QueryProductDetailsParams.newBuilder()
                .setProductList(buildProductList(LIFETIME_IDS, BillingClient.ProductType.INAPP))
                .build();

            billingClient.queryProductDetailsAsync(subParams, (subResult, subQueryResult) -> {
                billingClient.queryProductDetailsAsync(inappParams, (inappResult, inappQueryResult) -> {
                    cachedProducts.clear();
                    if (subResult.getResponseCode() != BillingClient.BillingResponseCode.OK) {
                        Log.w(TAG, "Subscription products query failed: " + subResult.getDebugMessage());
                    }
                    if (inappResult.getResponseCode() != BillingClient.BillingResponseCode.OK) {
                        Log.w(TAG, "In-app products query failed: " + inappResult.getDebugMessage());
                    }
                    List<ProductDetails> subDetails = productDetailsFrom(subQueryResult);
                    List<ProductDetails> inappDetails = productDetailsFrom(inappQueryResult);
                    if (subDetails != null) cachedProducts.addAll(subDetails);
                    if (inappDetails != null) cachedProducts.addAll(inappDetails);

                    JSArray products = new JSArray();
                    for (ProductDetails pd : cachedProducts) {
                        JSObject p = new JSObject();
                        p.put("id", pd.getProductId());
                        p.put("displayName", pd.getName());
                        p.put("description", pd.getDescription());

                        if (pd.getProductType().equals(BillingClient.ProductType.SUBS)) {
                            ProductDetails.PricingPhase phase = getDisplayPricingPhase(pd);
                            if (phase != null) {
                                p.put("displayPrice", phase.getFormattedPrice());
                                p.put("price", phase.getPriceAmountMicros() / 1_000_000.0);
                                p.put("currencyCode", phase.getPriceCurrencyCode());
                            }
                            p.put("type", "subscription");
                            p.put("period", periodForProduct(pd, phase));
                        } else {
                            ProductDetails.OneTimePurchaseOfferDetails otpd = pd.getOneTimePurchaseOfferDetails();
                            if (otpd != null) {
                                p.put("displayPrice", otpd.getFormattedPrice());
                                p.put("price", otpd.getPriceAmountMicros() / 1_000_000.0);
                                p.put("currencyCode", otpd.getPriceCurrencyCode());
                            }
                            p.put("type", "nonConsumable");
                            p.put("period", "lifetime");
                        }
                        products.put(p);
                    }

                    Log.i(TAG, "Loaded " + cachedProducts.size() + " products");
                    JSObject ret = new JSObject();
                    ret.put("products", products);
                    call.resolve(ret);
                });
            });
        }, () -> {
            JSObject ret = new JSObject();
            ret.put("products", new JSArray());
            ret.put("error", "billing_unavailable");
            call.resolve(ret);
        });
    }

    @PluginMethod
    public void purchase(PluginCall call) {
        String productId = call.getString("productId");
        if (productId == null) { call.reject("productId required"); return; }

        connectBilling(() -> {
            ProductDetails target = null;
            for (ProductDetails pd : cachedProducts) {
                if (pd.getProductId().equals(productId)) { target = pd; break; }
            }
            if (target != null) {
                launchPurchaseFlow(call, target);
            } else {
                // Product not cached — query on-demand before purchasing
                String productType = SUB_IDS.contains(productId)
                    ? BillingClient.ProductType.SUBS : BillingClient.ProductType.INAPP;
                QueryProductDetailsParams params = QueryProductDetailsParams.newBuilder()
                    .setProductList(Arrays.asList(
                        QueryProductDetailsParams.Product.newBuilder()
                            .setProductId(productId)
                            .setProductType(productType)
                            .build()))
                    .build();
                billingClient.queryProductDetailsAsync(params, (result, queryResult) -> {
                    List<ProductDetails> detailsList = productDetailsFrom(queryResult);
                    if (result.getResponseCode() == BillingClient.BillingResponseCode.OK
                            && detailsList != null && !detailsList.isEmpty()) {
                        ProductDetails fetched = detailsList.get(0);
                        cachedProducts.add(fetched);
                        launchPurchaseFlow(call, fetched);
                    } else {
                        call.reject("Product not found: " + productId);
                    }
                });
            }
        }, () -> call.reject("Billing unavailable"));
    }

    private ProductDetails.SubscriptionOfferDetails chooseOffer(ProductDetails target) {
        List<ProductDetails.SubscriptionOfferDetails> offers = target.getSubscriptionOfferDetails();
        if (offers == null || offers.isEmpty()) return null;

        String wantedPeriod = periodForProduct(target, null);
        ProductDetails.SubscriptionOfferDetails fallback = offers.get(0);
        for (ProductDetails.SubscriptionOfferDetails offer : offers) {
            List<ProductDetails.PricingPhase> phases = offer.getPricingPhases().getPricingPhaseList();
            if (phases == null || phases.isEmpty()) continue;
            for (ProductDetails.PricingPhase phase : phases) {
                String phasePeriod = periodForBillingPeriod(phase.getBillingPeriod());
                if (phase.getPriceAmountMicros() > 0 && phasePeriod.equals(wantedPeriod)) {
                    return offer;
                }
            }
        }
        return fallback;
    }

    private ProductDetails.PricingPhase getDisplayPricingPhase(ProductDetails target) {
        ProductDetails.SubscriptionOfferDetails offer = chooseOffer(target);
        if (offer == null || offer.getPricingPhases() == null) return null;
        List<ProductDetails.PricingPhase> phases = offer.getPricingPhases().getPricingPhaseList();
        if (phases == null || phases.isEmpty()) return null;

        String wantedPeriod = periodForProduct(target, null);
        ProductDetails.PricingPhase fallbackPaid = null;
        ProductDetails.PricingPhase lastPhase = phases.get(phases.size() - 1);
        for (ProductDetails.PricingPhase phase : phases) {
            if (phase.getPriceAmountMicros() > 0) {
                fallbackPaid = phase;
                if (periodForBillingPeriod(phase.getBillingPeriod()).equals(wantedPeriod)) {
                    return phase;
                }
            }
        }
        return fallbackPaid != null ? fallbackPaid : lastPhase;
    }

    private String periodForProduct(ProductDetails product, ProductDetails.PricingPhase phase) {
        String id = product.getProductId();
        if (id != null) {
            if (id.contains("lifetime")) return "lifetime";
            if (id.contains("annual") || id.contains("year")) return "annual";
            if (id.contains("monthly") || id.contains("month")) return "monthly";
        }
        if (phase != null) return periodForBillingPeriod(phase.getBillingPeriod());
        return "monthly";
    }

    private String periodForBillingPeriod(String billingPeriod) {
        if (billingPeriod == null) return "monthly";
        return billingPeriod.contains("Y") ? "annual" : "monthly";
    }

    private boolean isRecognizedPurchase(Purchase purchase, boolean lifetimeOnly) {
        if (purchase == null || purchase.getPurchaseState() != Purchase.PurchaseState.PURCHASED) return false;
        if (purchase.isSuspended()) return false;
        for (String pid : purchase.getProducts()) {
            if (lifetimeOnly) {
                if (LIFETIME_IDS.contains(pid)) return true;
            } else if (PRODUCT_IDS.contains(pid)) {
                return true;
            }
        }
        return false;
    }

    private void launchPurchaseFlow(PluginCall call, ProductDetails target) {
        Activity activity = getActivity();
        if (activity == null) { call.reject("Activity not available"); return; }

        pendingPurchaseCall = call;

        BillingFlowParams.ProductDetailsParams.Builder pdpBuilder =
            BillingFlowParams.ProductDetailsParams.newBuilder().setProductDetails(target);

        if (target.getProductType().equals(BillingClient.ProductType.SUBS)) {
            ProductDetails.SubscriptionOfferDetails offer = chooseOffer(target);
            if (offer != null) {
                pdpBuilder.setOfferToken(offer.getOfferToken());
            }
        }

        BillingFlowParams flowParams = BillingFlowParams.newBuilder()
            .setProductDetailsParamsList(Arrays.asList(pdpBuilder.build()))
            .build();

        BillingResult flowResult = billingClient.launchBillingFlow(activity, flowParams);
        if (flowResult.getResponseCode() != BillingClient.BillingResponseCode.OK) {
            pendingPurchaseCall = null;
            call.reject("Failed to launch billing: " + flowResult.getDebugMessage());
        }
    }

    @Override
    public void onPurchasesUpdated(@NonNull BillingResult result, List<Purchase> purchases) {
        if (result.getResponseCode() == BillingClient.BillingResponseCode.OK && purchases != null) {
            boolean foundPremium = false;
            for (Purchase purchase : purchases) {
                if (isRecognizedPurchase(purchase, false)) {
                    foundPremium = true;
                    acknowledgePurchase(purchase);
                }
            }
            setPremium(foundPremium);
            if (pendingPurchaseCall != null) {
                JSObject ret = new JSObject();
                ret.put("success", foundPremium);
                ret.put("isPremium", foundPremium);
                pendingPurchaseCall.resolve(ret);
                pendingPurchaseCall = null;
            }
        } else if (result.getResponseCode() == BillingClient.BillingResponseCode.USER_CANCELED) {
            if (pendingPurchaseCall != null) {
                JSObject ret = new JSObject();
                ret.put("success", false);
                ret.put("cancelled", true);
                ret.put("isPremium", false);
                pendingPurchaseCall.resolve(ret);
                pendingPurchaseCall = null;
            }
        } else {
            if (pendingPurchaseCall != null) {
                pendingPurchaseCall.reject("Purchase failed: " + result.getDebugMessage());
                pendingPurchaseCall = null;
            }
        }
    }

    private void acknowledgePurchase(Purchase purchase) {
        if (purchase.getPurchaseState() == Purchase.PurchaseState.PURCHASED && !purchase.isAcknowledged()) {
            AcknowledgePurchaseParams params = AcknowledgePurchaseParams.newBuilder()
                .setPurchaseToken(purchase.getPurchaseToken())
                .build();
            billingClient.acknowledgePurchase(params, result ->
                Log.i(TAG, "Acknowledge: " + result.getResponseCode()));
        }
    }

    @PluginMethod
    public void restore(PluginCall call) {
        connectBilling(() -> {
            boolean[] found = {false};

            // Check subscriptions
            billingClient.queryPurchasesAsync(
                QueryPurchasesParams.newBuilder().setProductType(BillingClient.ProductType.SUBS).build(),
                (subResult, subPurchases) -> {
                    if (subPurchases != null) {
                        for (Purchase p : subPurchases) {
                            if (isRecognizedPurchase(p, false)) {
                                found[0] = true;
                                acknowledgePurchase(p);
                                break;
                            }
                        }
                    }

                    // Check in-app (lifetime)
                    billingClient.queryPurchasesAsync(
                        QueryPurchasesParams.newBuilder().setProductType(BillingClient.ProductType.INAPP).build(),
                        (inappResult, inappPurchases) -> {
                            if (inappPurchases != null) {
                                for (Purchase p : inappPurchases) {
                                    if (isRecognizedPurchase(p, true)) {
                                        found[0] = true;
                                        acknowledgePurchase(p);
                                        break;
                                    }
                                }
                            }

                            setPremium(found[0]);
                            Log.i(TAG, "Restore: isPremium=" + found[0]);
                            JSObject ret = new JSObject();
                            ret.put("isPremium", found[0]);
                            call.resolve(ret);
                        });
                });
        }, () -> {
            setPremium(false);
            JSObject ret = new JSObject();
            ret.put("isPremium", false);
            ret.put("error", "billing_unavailable");
            call.resolve(ret);
        });
    }

    @PluginMethod
    public void getEntitlements(PluginCall call) {
        restore(call); // Same logic — check active purchases
    }

    private void setPremium(boolean premium) {
        SharedPreferences prefs = getContext().getSharedPreferences("obubba_prefs", android.content.Context.MODE_PRIVATE);
        prefs.edit().putBoolean("ob_premium", premium).apply();
    }

    private List<QueryProductDetailsParams.Product> buildProductList(List<String> ids, String type) {
        List<QueryProductDetailsParams.Product> list = new ArrayList<>();
        for (String id : ids) {
            list.add(QueryProductDetailsParams.Product.newBuilder()
                .setProductId(id)
                .setProductType(type)
                .build());
        }
        return list;
    }

    private List<ProductDetails> productDetailsFrom(QueryProductDetailsResult queryResult) {
        if (queryResult == null || queryResult.getProductDetailsList() == null) return new ArrayList<>();
        return queryResult.getProductDetailsList();
    }
}

File: android/app/src/main/java/com/obubba/app/plugins/StorePlugin.kt
package com.obubba.app.plugins

import android.app.Activity
import android.util.Log
import com.android.billingclient.api.*
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

/**
 * Capacitor plugin for Google Play Billing (In-App Subscriptions & Purchases).
 * Mirrors the iOS StorePlugin interface: getProducts, purchase, restore, getEntitlements.
 * Uses Google Play Billing Library v7.
 */
@CapacitorPlugin(name = "OBStore")
class StorePlugin : Plugin(), PurchasesUpdatedListener {

    companion object {
        private const val TAG = "OBStore"
    }

    // Product IDs -- must match Google Play Console
    private val productIds = listOf(
        "com.obubba.premium.monthly",
        "com.obubba.premium.annual",
        "com.obubba.premium.lifetime"
    )

    // Subscription product IDs (auto-renewable)
    private val subscriptionIds = setOf(
        "com.obubba.premium.monthly",
        "com.obubba.premium.annual"
    )

    // In-app (one-time) product IDs
    private val inAppIds = setOf(
        "com.obubba.premium.lifetime"
    )

    private lateinit var billingClient: BillingClient
    private var pendingPurchaseCall: PluginCall? = null

    // Cache product details after loading
    private var cachedProductDetails: MutableMap<String, ProductDetails> = mutableMapOf()

    override fun load() {
        billingClient = BillingClient.newBuilder(context)
            .setListener(this)
            .enablePendingPurchases()
            .build()

        // Connect eagerly so products are ready when needed
        ensureConnected {}
    }

    // ── Connection helper ──────────────────────────────────────────

    private fun ensureConnected(onReady: () -> Unit) {
        if (billingClient.isReady) {
            onReady()
            return
        }
        billingClient.startConnection(object : BillingClientStateListener {
            override fun onBillingSetupFinished(result: BillingResult) {
                if (result.responseCode == BillingClient.BillingResponseCode.OK) {
                    Log.d(TAG, "Billing client connected")
                    onReady()
                } else {
                    Log.e(TAG, "Billing setup failed: ${result.debugMessage}")
                }
            }

            override fun onBillingServiceDisconnected() {
                Log.w(TAG, "Billing service disconnected")
            }
        })
    }

    // ── getProducts ────────────────────────────────────────────────
    // Returns product info with prices, matching iOS format:
    // { products: [{ id, displayName, description, displayPrice, price, type, period, periodValue }] }

    @PluginMethod
    fun getProducts(call: PluginCall) {
        ensureConnected {
            queryAllProducts(call)
        }
    }

    private fun queryAllProducts(call: PluginCall) {
        val allProducts = mutableListOf<JSObject>()
        var queriesRemaining = 2 // subs + inapp

        // Query subscriptions
        val subParams = QueryProductDetailsParams.newBuilder()
            .setProductList(
                subscriptionIds.map { id ->
                    QueryProductDetailsParams.Product.newBuilder()
                        .setProductId(id)
                        .setProductType(BillingClient.ProductType.SUBS)
                        .build()
                }
            )
            .build()

        billingClient.queryProductDetailsAsync(subParams) { result, detailsList ->
            if (result.responseCode == BillingClient.BillingResponseCode.OK && detailsList != null) {
                for (details in detailsList) {
                    cachedProductDetails[details.productId] = details
                    allProducts.add(productDetailsToJS(details))
                }
            }
            synchronized(this) {
                queriesRemaining--
                if (queriesRemaining == 0) {
                    resolveProducts(call, allProducts)
                }
            }
        }

        // Query in-app (lifetime)
        val inAppParams = QueryProductDetailsParams.newBuilder()
            .setProductList(
                inAppIds.map { id ->
                    QueryProductDetailsParams.Product.newBuilder()
                        .setProductId(id)
                        .setProductType(BillingClient.ProductType.INAPP)
                        .build()
                }
            )
            .build()

        billingClient.queryProductDetailsAsync(inAppParams) { result, detailsList ->
            if (result.responseCode == BillingClient.BillingResponseCode.OK && detailsList != null) {
                for (details in detailsList) {
                    cachedProductDetails[details.productId] = details
                    allProducts.add(productDetailsToJS(details))
                }
            }
            synchronized(this) {
                queriesRemaining--
                if (queriesRemaining == 0) {
                    resolveProducts(call, allProducts)
                }
            }
        }
    }

    private fun resolveProducts(call: PluginCall, products: List<JSObject>) {
        val ret = JSObject()
        val arr = JSArray()
        for (p in products) {
            arr.put(p)
        }
        ret.put("products", arr)
        Log.d(TAG, "Loaded ${products.size} products")
        call.resolve(ret)
    }

    private fun productDetailsToJS(details: ProductDetails): JSObject {
        val obj = JSObject()
        obj.put("id", details.productId)
        obj.put("displayName", details.name)
        obj.put("description", details.description)

        if (details.productType == BillingClient.ProductType.SUBS) {
            obj.put("type", "subscription")
            // Use the first pricing phase of the first offer (base plan)
            val offer = details.subscriptionOfferDetails?.firstOrNull()
            val pricingPhase = offer?.pricingPhases?.pricingPhaseList?.firstOrNull()
            if (pricingPhase != null) {
                obj.put("displayPrice", pricingPhase.formattedPrice)
                obj.put("price", pricingPhase.priceAmountMicros / 1_000_000.0)
                // Parse billing period (ISO 8601 duration: P1M, P1Y, etc.)
                val period = pricingPhase.billingPeriod
                when {
                    period.contains("Y") -> {
                        obj.put("period", "annual")
                        obj.put("periodValue", 1)
                    }
                    period.contains("M") -> {
                        obj.put("period", "monthly")
                        obj.put("periodValue", 1)
                    }
                    else -> {
                        obj.put("period", "other")
                        obj.put("periodValue", 1)
                    }
                }
            }
        } else {
            // In-app product (lifetime)
            obj.put("type", "nonConsumable")
            val oneTimePurchaseOfferDetails = details.oneTimePurchaseOfferDetails
            if (oneTimePurchaseOfferDetails != null) {
                obj.put("displayPrice", oneTimePurchaseOfferDetails.formattedPrice)
                obj.put("price", oneTimePurchaseOfferDetails.priceAmountMicros / 1_000_000.0)
            }
        }

        return obj
    }

    // ── purchase ───────────────────────────────────────────────────
    // Expects { productId: "com.obubba.premium.monthly" }
    // Returns { success, productId, isPremium } or { success: false, cancelled: true }

    @PluginMethod
    fun purchase(call: PluginCall) {
        val productId = call.getString("productId")
        if (productId == null) {
            call.reject("productId is required")
            return
        }

        ensureConnected {
            launchPurchaseFlow(call, productId)
        }
    }

    private fun launchPurchaseFlow(call: PluginCall, productId: String) {
        val details = cachedProductDetails[productId]
        if (details == null) {
            // Product not cached yet -- query first
            val productType = if (subscriptionIds.contains(productId))
                BillingClient.ProductType.SUBS else BillingClient.ProductType.INAPP

            val params = QueryProductDetailsParams.newBuilder()
                .setProductList(
                    listOf(
                        QueryProductDetailsParams.Product.newBuilder()
                            .setProductId(productId)
                            .setProductType(productType)
                            .build()
                    )
                )
                .build()

            billingClient.queryProductDetailsAsync(params) { result, detailsList ->
                if (result.responseCode == BillingClient.BillingResponseCode.OK && detailsList != null) {
                    val fetched = detailsList.firstOrNull()
                    if (fetched != null) {
                        cachedProductDetails[productId] = fetched
                        startBillingFlow(call, fetched)
                    } else {
                        call.reject("Product not found: $productId")
                    }
                } else {
                    call.reject("Failed to query product: ${result.debugMessage}")
                }
            }
        } else {
            startBillingFlow(call, details)
        }
    }

    private fun startBillingFlow(call: PluginCall, details: ProductDetails) {
        val activity = activity
        if (activity == null) {
            call.reject("Activity not available")
            return
        }

        pendingPurchaseCall = call

        val productDetailsParamsBuilder = BillingFlowParams.ProductDetailsParams.newBuilder()
            .setProductDetails(details)

        // For subscriptions, we must set the offer token
        if (details.productType == BillingClient.ProductType.SUBS) {
            val offerToken = details.subscriptionOfferDetails?.firstOrNull()?.offerToken
            if (offerToken != null) {
                productDetailsParamsBuilder.setOfferToken(offerToken)
            }
        }

        val flowParams = BillingFlowParams.newBuilder()
            .setProductDetailsParamsList(listOf(productDetailsParamsBuilder.build()))
            .build()

        val result = billingClient.launchBillingFlow(activity, flowParams)
        if (result.responseCode != BillingClient.BillingResponseCode.OK) {
            pendingPurchaseCall = null
            call.reject("Failed to launch billing flow: ${result.debugMessage}")
        }
    }

    // ── PurchasesUpdatedListener callback ──────────────────────────

    override fun onPurchasesUpdated(result: BillingResult, purchases: MutableList<Purchase>?) {
        val call = pendingPurchaseCall
        pendingPurchaseCall = null

        when (result.responseCode) {
            BillingClient.BillingResponseCode.OK -> {
                if (purchases != null && purchases.isNotEmpty()) {
                    // Acknowledge/consume all purchases
                    for (purchase in purchases) {
                        handlePurchase(purchase, call)
                    }
                } else {
                    call?.resolve(JSObject().apply {
                        put("success", false)
                        put("isPremium", false)
                    })
                }
            }
            BillingClient.BillingResponseCode.USER_CANCELED -> {
                Log.d(TAG, "Purchase cancelled by user")
                call?.resolve(JSObject().apply {
                    put("success", false)
                    put("cancelled", true)
                    put("isPremium", false)
                })
            }
            BillingClient.BillingResponseCode.ITEM_ALREADY_OWNED -> {
                Log.d(TAG, "Item already owned")
                setPremiumFlag(true)
                call?.resolve(JSObject().apply {
                    put("success", true)
                    put("isPremium", true)
                })
            }
            else -> {
                Log.e(TAG, "Purchase error: ${result.responseCode} - ${result.debugMessage}")
                call?.reject("Purchase failed: ${result.debugMessage}")
            }
        }
    }

    private fun handlePurchase(purchase: Purchase, call: PluginCall?) {
        if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED) {
            // Acknowledge the purchase if not already acknowledged
            if (!purchase.isAcknowledged) {
                val ackParams = AcknowledgePurchaseParams.newBuilder()
                    .setPurchaseToken(purchase.purchaseToken)
                    .build()

                billingClient.acknowledgePurchase(ackParams) { ackResult ->
                    if (ackResult.responseCode == BillingClient.BillingResponseCode.OK) {
                        Log.d(TAG, "Purchase acknowledged: ${purchase.products}")
                        setPremiumFlag(true)
                        call?.resolve(JSObject().apply {
                            put("success", true)
                            put("productId", purchase.products.firstOrNull() ?: "")
                            put("isPremium", true)
                        })
                    } else {
                        Log.e(TAG, "Acknowledge failed: ${ackResult.debugMessage}")
                        // Still grant premium -- acknowledgement can be retried
                        setPremiumFlag(true)
                        call?.resolve(JSObject().apply {
                            put("success", true)
                            put("productId", purchase.products.firstOrNull() ?: "")
                            put("isPremium", true)
                        })
                    }
                }
            } else {
                setPremiumFlag(true)
                call?.resolve(JSObject().apply {
                    put("success", true)
                    put("productId", purchase.products.firstOrNull() ?: "")
                    put("isPremium", true)
                })
            }
        } else if (purchase.purchaseState == Purchase.PurchaseState.PENDING) {
            Log.d(TAG, "Purchase pending")
            call?.resolve(JSObject().apply {
                put("success", false)
                put("pending", true)
                put("isPremium", false)
            })
        } else {
            call?.resolve(JSObject().apply {
                put("success", false)
                put("isPremium", false)
            })
        }
    }

    // ── restore ────────────────────────────────────────────────────
    // Queries Google Play for existing purchases. Returns { isPremium: bool }

    @PluginMethod
    fun restore(call: PluginCall) {
        ensureConnected {
            checkAllPurchases { hasPremium ->
                setPremiumFlag(hasPremium)
                Log.d(TAG, "Restore complete, isPremium: $hasPremium")
                call.resolve(JSObject().apply {
                    put("isPremium", hasPremium)
                })
            }
        }
    }

    // ── getEntitlements ────────────────────────────────────────────
    // Check current active entitlements. Returns { isPremium: bool }

    @PluginMethod
    fun getEntitlements(call: PluginCall) {
        ensureConnected {
            checkAllPurchases { hasPremium ->
                setPremiumFlag(hasPremium)
                call.resolve(JSObject().apply {
                    put("isPremium", hasPremium)
                })
            }
        }
    }

    // ── Internal helpers ───────────────────────────────────────────

    /**
     * Check both subscription and in-app purchases for any active premium entitlement.
     */
    private fun checkAllPurchases(callback: (Boolean) -> Unit) {
        var hasPremium = false
        var queriesRemaining = 2

        // Check subscriptions
        val subParams = QueryPurchasesParams.newBuilder()
            .setProductType(BillingClient.ProductType.SUBS)
            .build()

        billingClient.queryPurchasesAsync(subParams) { result, purchases ->
            if (result.responseCode == BillingClient.BillingResponseCode.OK) {
                for (purchase in purchases) {
                    if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED) {
                        for (product in purchase.products) {
                            if (productIds.contains(product)) {
                                hasPremium = true
                                // Acknowledge if needed
                                acknowledgePurchaseIfNeeded(purchase)
                            }
                        }
                    }
                }
            }
            synchronized(this) {
                queriesRemaining--
                if (queriesRemaining == 0) {
                    callback(hasPremium)
                }
            }
        }

        // Check in-app purchases (lifetime)
        val inAppParams = QueryPurchasesParams.newBuilder()
            .setProductType(BillingClient.ProductType.INAPP)
            .build()

        billingClient.queryPurchasesAsync(inAppParams) { result, purchases ->
            if (result.responseCode == BillingClient.BillingResponseCode.OK) {
                for (purchase in purchases) {
                    if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED) {
                        for (product in purchase.products) {
                            if (productIds.contains(product)) {
                                hasPremium = true
                                acknowledgePurchaseIfNeeded(purchase)
                            }
                        }
                    }
                }
            }
            synchronized(this) {
                queriesRemaining--
                if (queriesRemaining == 0) {
                    callback(hasPremium)
                }
            }
        }
    }

    private fun acknowledgePurchaseIfNeeded(purchase: Purchase) {
        if (!purchase.isAcknowledged && purchase.purchaseState == Purchase.PurchaseState.PURCHASED) {
            val ackParams = AcknowledgePurchaseParams.newBuilder()
                .setPurchaseToken(purchase.purchaseToken)
                .build()
            billingClient.acknowledgePurchase(ackParams) { result ->
                if (result.responseCode == BillingClient.BillingResponseCode.OK) {
                    Log.d(TAG, "Retroactively acknowledged: ${purchase.products}")
                } else {
                    Log.w(TAG, "Failed to acknowledge: ${result.debugMessage}")
                }
            }
        }
    }

    /**
     * Persist premium status to SharedPreferences so it survives restarts.
     */
    private fun setPremiumFlag(isPremium: Boolean) {
        context.getSharedPreferences("obubba_prefs", android.content.Context.MODE_PRIVATE)
            .edit()
            .putBoolean("ob_premium", isPremium)
            .apply()
    }

    override fun handleOnDestroy() {
        super.handleOnDestroy()
        if (::billingClient.isInitialized) {
            billingClient.endConnection()
        }
    }
}
